valhalla/thor/dijkstras.h

```diff
@@ -0,0 +1,273 @@
+   */
+  Dijkstras();
+
+  /**
+   * Destructor
+   */
+  virtual ~Dijkstras();
+
+  /**
+   * Clear the temporary memory (adjacency list, edgestatus, edgelabels)
+   */
+  void Clear();
+
+  /**
+   * Compute the best first graph traversal from a list of origin locations
+   * @param  origin_locs  List of origin locations.
+   * @param  graphreader  Graphreader
+   * @param  mode_costing List of costing objects
+   * @param  mode         Travel mode
+   */
+  void Compute(google::protobuf::RepeatedPtrField<valhalla::Location>& origin_locs,
+               baldr::GraphReader& graphreader,
+               const std::shared_ptr<sif::DynamicCost>* mode_costing,
+               const sif::TravelMode mode);
+
+  /**
+   * Compute the best first graph traversal to a list of destination locations
+   * @param  origin_locs  List of origin locations.
+   * @param  graphreader  Graphreader
+   * @param  mode_costing List of costing objects
+   * @param  mode         Travel mode
+   */
+  void ComputeReverse(google::protobuf::RepeatedPtrField<valhalla::Location>& dest_locations,
+                      baldr::GraphReader& graphreader,
+                      const std::shared_ptr<sif::DynamicCost>* mode_costing,
+                      const sif::TravelMode mode);
+
+  /**
+   * Compute the best first graph traversal from a list of origin locations using multimodal
+   * @param  origin_locs  List of origin locations.
+   * @param  graphreader  Graphreader
+   * @param  mode_costing List of costing objects
+   * @param  mode         Travel mode
+   */
+  void ComputeMultiModal(google::protobuf::RepeatedPtrField<valhalla::Location>& origin_locations,
+                         baldr::GraphReader& graphreader,
+                         const std::shared_ptr<sif::DynamicCost>* mode_costing,
+                         const sif::TravelMode mode);
+
+protected:
+  // Virtual function called when expanding a node
+  //
+  // Children can implement this to customize behaviour
+  virtual void ExpandingNode(baldr::GraphReader& graphreader,
+                             const sif::EdgeLabel& pred,
+                             const ExpandingNodeMiscInfo& info){};
+
+  // A child-class can implement this to configure how the graph expands
+  virtual RouteCallbackRecommendedAction RouteCallbackDecideAction(baldr::GraphReader& graphreader,
+                                                                   const sif::EdgeLabel& pred,
+                                                                   const InfoRoutingType route_type) {
+    return RouteCallbackRecommendedAction::no_action;
+  };
+
+  bool has_date_time_;
+  int start_tz_index_;   // Timezone at the start of the Dijkstras
+  sif::TravelMode mode_; // Current travel mode
+  uint32_t access_mode_; // Access mode used by the costing method
+
+  // For multimodal
+  bool date_set_;
+  bool date_before_tile_;
+  uint32_t date_;
+  uint32_t dow_;
+  uint32_t day_;
+  uint32_t max_transfer_distance_;
```

Comment:
With the recent overflow/underflow bugs that we came across, do we want to reconsider using `unsigned` types for quantities (eg distance here)?

Reply:
I'm open to ideas. Is there a risk of hiding underflow bugs from being detected if we use signed numbers?

Reply:
> I'm open to ideas. Is there a risk of hiding underflow bugs from being detected if we use signed numbers?

I'm not sure I follow. Won't signed integer show the underflow (instead if it was unsigned, it would just wraparound to a high value)?

Reply:
Discussed in person last Friday. Essentially signed numbers would be combined with lots of asserts checking for negative numbers.